refactor: rewrite tests to use webpack compiler, drop node 8

BeeeQueue opened the pull request:
- Replaces the `{}` replacement with `"MISSING_ENV_VAR"` which seems to not error, which the old one started doing for some reason
- Uses the [Webpack target](https://webpack.js.org/configuration/target) to decide to stub `process.env` or not.
  _For now it will not do it for Node and Electron targets_
- Rewrite all the tests to use the webpack compiler, and then testing on the output files
- Drops support for Node 8 as the tests can't run on it because of `BigInt` not existing

You can test it for yourselves by using a yarn resolution:
```json
  "resolutions": {
    "dotenv-webpack": "npm:@beequeue/dotenv-webpack"
  },
```

Fixes #271
Closes #289 

CI runs: https://github.com/BeeeQueue/dotenv-webpack/actions?query=branch%3Afix-stubbing

sapkra commented:
With this solution the code for setters will still break. But like I already said in the issue you can just omit it for node environments.

Browser:
- cannot have env var setters
- don't have process.env - need to be replaced by `{}`

Node/Electron:
- can have setters
- has process.env  - no need to replace it

Maybe I'm wrong but it should work like this.

sapkra commented on src/index.js:
```diff
@@ -144,8 +150,14 @@ class Dotenv {
+    // https://github.com/mrsteele/dotenv-webpack/issues/240#issuecomment-710231534
+    // However, if someone targets Node or Electron `process.env` still exists, and should therefore be kept
+    // https://webpack.js.org/configuration/target
+    if (!target.startsWith('node') && !target.startsWith('electron')) {
```
Not all `electron` targets are node targets. Some of them are for the browser environment.

sapkra commented on src/index.js:
```diff
@@ -144,8 +150,14 @@ class Dotenv {
+    // https://webpack.js.org/configuration/target
+    if (!target.startsWith('node') && !target.startsWith('electron')) {
+      // Results in `"MISSING_ENV_VAR".NAME` which is valid JS
+      formatted['process.env'] = '"MISSING_ENV_VAR"'
```
If you already omitting it for node and electron why is there a need for changing it to `"MISSING_ENV_VAR"`? We could still use `{}` because it was only breaking for setters which are not existing in browser environments.

BeeeQueue replied:
It is erroring even when just accessing `{}.NAME` for us. We had to update our own stub to this, as well as opening this PR or it will give us the famous `Unexpected character '.'`.

I checked and it is indeed just accessing `{}` that is breaking, and specifically only when this plugin does its replacement.

sapkra replied:
Interesting, you're right. Well, then we have to use this workaround I guess.

Another solution would be the following but it will not make any different to yours:

```javascript
formatted['process.env'] = 'new Object({})'
```

BeeeQueue replied:
I like this string because it also does explain why we have this weird code left over. Rather than `{}.NAME` it literally says that an environment variable is missing

mrsteele commented:
Thanks for checking this out @BeeeQueue, any chance you ran this with all the supported versions of Webpack? Would love to keep it compatible with the other versions. Also, any chance we can add some tests in for this new check?

BeeeQueue commented:
Turns out that the CI actions are not being run on PRs, and all the tests are failing! 😄 

For this solution to be testable I think we need to set up a webpack compiler and compile stuff in the tests, so I'll have to do that.

mrsteele commented:
As long as we have at least 4 and 5 I think I'd be good. I imagine supporting those older versions would be challenging?

mrsteele commented:
(I went back and looked and webpack@v3 was last updated May 2018)

BeeeQueue commented:
More about it being time consuming to (manually!) test for them, and then creating workarounds if something is found when no one should really be using them.

mrsteele commented:
@BeeeQueue would it make more sense to make the user make that distinction? https://github.com/mrsteele/dotenv-webpack/pull/289 <- That PR puts a flag in the config to opt out of the stub but curious on your thoughts.

BeeeQueue commented:
I think having the option to override this functionality is a very good idea, and having sensible defaults on when to do it ourselves.

Or maybe it would be even better to extract that to another plugin, `ProcessEnvStubPlugin` that will take care of this...

Another benefit of merging this PR is that the tests will be better, as in they are testing the real behavior of webpack as the tests will use the compiler now

BeeeQueue commented:
Since the test refactoring is very large I'm changing this PR to be about the tests and dropping Node only, and opening another one for the fixes afterwards.

mrsteele reviewed (commented):
This is looking really good. Any chance we can add the stub functionality in as well as drop the deprecated argument? I think once those are in this is a perfect release candidate for #363

BeeeQueue commented:
Like I said I can do that in a PR after this one is merged, just so it's easier to develop/review/diff

mrsteele commented:
👍 Cool I'll pull it down and run a test during my lunch here. If all looks good we will have this in within the hour.

mrsteele commented:
Actually, can we at least remove the deprecated statement? That would be a breaking change and so is this and I'd rather only do one breaking change in a day. You could also remove the test for it which would clean up more things.

BeeeQueue commented:
Is the project set up to publish on each commit? If so I'll just put everything in here

mrsteele commented:
@BeeeQueue Yes, semantic release. I was going to squash it and add the breaking change commit message.

BeeeQueue commented:
Everything should be in now.

mrsteele commented:
@BeeeQueue where did we land on https://github.com/mrsteele/dotenv-webpack/pull/360#issuecomment-790448213?

Manually configure to stub, or automate based on webpack config?

BeeeQueue commented:
As of now it's automated based on the target.

mrsteele commented:
What do you think @sapkra?
1. Implicitly defined from the webpack config: https://github.com/mrsteele/dotenv-webpack/pull/360/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R20
2. Explicitly defined from the Dotenv config: https://github.com/mrsteele/dotenv-webpack/pull/289/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R148
   - upgrading to support the `MISSING_ENV_VAR` of course

BeeeQueue commented:
I can add the option in this PR as well, would make sense to have an override for the automatic behavior.

sapkra commented:
I like the implicit approach much more because it should automatically work for all users without telling them to manually configure it. But of course both approaches could be combined.

@BeeeQueue wrote the same while I was typing this comment.

mrsteele commented:
Love it! Best of both worlds! Lets do both, by default we would assume but of course if something is a little more complex we can override manually.

BeeeQueue commented on src/index.js:
```diff
@@ -144,8 +140,20 @@ class Dotenv {
+    // However, if someone targets Node or Electron `process.env` still exists, and should therefore be kept
+    // https://webpack.js.org/configuration/target
+    if (
+      this.config.ignoreStub !== true &&
+      (
+        this.config.ignoreStub === false ||
+        (!target.startsWith('node') && !isMainThreadElectron(target))
+      )
```
This logic became pretty ugly now, but it is explicitly tested so it should™️ be fine.

TL;DR: Use config if it's passed, otherwise use the target logic

BeeeQueue replied:
Codeclimate is complaining about it being too complex due to many lines, but it has many lines to make it more readable and less complex 😖

mrsteele commented:
Pulling things down and so far things are looking fine, but I realized a funny edge-case: Should we even open up the polyfill concept when your version of webpack  less than 5?

BeeeQueue commented:
I don't think it will break anything if we do the replacements in browsers since `process.env` is just an empty object there, but I can check if there's an easy way to get the webpack version in the plugin

mrsteele commented:
Yeah, might want to pull out the value interpretation part as its own thing.

Another thing looking over this, we only target electron main, should we also account for node or node-like envs?

https://webpack.js.org/configuration/target/

BeeeQueue commented:
I extracted the check to its own function, made it not run on webpack <5, and changed the check so it will not stub if the target _includes_ `node`, rather than starts with it.

I have no idea what the various other node targets are used for or if they have `process`, but if something breaks for the few people who use those I'm sure they'll come screaming...

mrsteele commented:
Everything I have tested so far looks good, seems to be right where it needs to be. Really great work and excited to get this out there!

On the browser list... target is SO confusing and not seeing a lot of documentation about the inner workings. 

Looking at the [source-code](https://github.com/webpack/webpack/blob/master/lib/config/target.js) is a little more illuminating.

Would it make more sense to use `getTargetsProperties` and if `!node` then we stub `MISSING_ENV_VAR` since that is ultimately the problem?

BeeeQueue commented:
I think the current checks are good enough.

`browserslist` is a special case that we can't account for, since it can target whatever it wants (`chrome 89`, `node 14`, `es5`), but it's mainly used for browsers so we can assume it as browser for now.

mrsteele commented:
At minimum `node` would need to be added, and I think it just begs the question on what others should be added. I personally don't know enough about the other targets to make that decision.

Do you think it would be safer to just drop the implicit route so we don't overstep some of those grey-area targets?

BeeeQueue commented:
As I said I think the current implementation will work for 90% of users, if not more.

If someone has a very special setup they are most likely an advanced user and should be able to debug and find our solution and how to disable it if they need to.

And I couldn't find any info on if `getTargetsProperties` is a public api

BeeeQueue commented on test/index.test.js:
```diff
@@ -82,7 +83,7 @@ beforeEach(() => {
   const outputDir = resolve(__dirname, `output/${hash(expect.getState().currentTestName)}`)
   try {
-    rmdirSync(outputDir, { recursive: true })
+    removeSync(outputDir)
```
Sadly `recursive: true` was added in Node 12, and as such our [Node 10 tests fail](https://github.com/BeeeQueue/dotenv-webpack/runs/2040389313?check_suite_focus=true) if we use it.

mrsteele commented:
Looking really good. I'm about ready to push this button. Anything else left?

BeeeQueue commented:
Not that I can think of. I'm sure there'll be something after it's out though. That's just how it is 😄
